modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/http/conn/ProxyConfig.java

```diff
@@ -85,6 +116,196 @@ public HttpHost selectProxy(final HttpHost target) {
+            return knownProxyConfigMap.get(targetHost);
+        }
+
+        if (knownDirectHosts.contains(targetHost)) {
+            return null;
+        }
+
+        boolean defaultProfile = false;
+        for (String key : proxyProfileMap.keySet()) {
+            if (key.equals("*")) {
+                defaultProfile = true;
+                continue;
+            }
+            if (targetHost.matches(key)) {
+                return getProxyProfileConfig(targetHost, key);
+            }
+        }
+
+        if (defaultProfile) {
+            return getProxyProfileConfig(targetHost, "*");
+        }
+
+        return null;
+    }
+
+    /**
+     * Selects the proxyProfile for the given key and gets the bypass set. Matches the targetHost against the
+     * bypass set. If it is matched then adds the targetHost to the knownDirectHosts List and returns null.
+     * Otherwise (i.e the targetHost is not matched in the bypass proxy) puts the proxyProfile against the targetHost
+     * into the knownProxyConfigMap and returns the proxyProfileConfig
+     *
+     * @param targetHost request's targeted host
+     * @param key proxyProfileMap's key, if default profile then the key is "*"
+     * @return proxyProfileConfig
+     */
+    private ProxyProfileConfig getProxyProfileConfig(String targetHost, String key) {
+        ProxyProfileConfig proxyProfileConfig = proxyProfileMap.get(key);
+        Set<String> proxyByPass = proxyProfileConfig.getProxyByPass();
+        for (String bypass : proxyByPass) {
+            if (targetHost.matches(bypass)) {
+                knownDirectHosts.add(targetHost);
+                return null;
+            }
+        }
+        knownProxyConfigMap.put(targetHost, proxyProfileConfig);
+        return proxyProfileConfig;
+    }
+
+
+    /**
+     * select the proxy credential for the targetHost
+     *
+     * @param targetHost targeted host
+     * @return proxy credential for the given end point, if not returns null
+     */
+    public UsernamePasswordCredentials getCredentialsForTargetHost(String targetHost) {
+
+        UsernamePasswordCredentials credentials = null;
+        ProxyProfileConfig proxyProfileForTargetHost = getProxyProfileForTargetHost(targetHost);
+
+        if (proxyProfileForTargetHost != null) {
+            credentials = proxyProfileForTargetHost.getCredentials();
+        }
+
+        return credentials;
+    }
+
+    /**
+     * returns appropriate log message based on the proxy configuration
+     * whether loading proxy profile or single proxy server or no proxy configured
+     *
+     * @return log message
+     */
+    public String logProxyConfig() {
+        if (isProxyProfileConfigured()) {
+            return "HTTP Sender using proxy profile";
+        }
+
+        if (this.proxy != null) {
+            return "HTTP Sender using Proxy " + getProxy() + " and  bypassing " + getProxyBypass();
+        } else {
+            return "No proxy configuration found";
+        }
```

Comment:
Hi Buddhima, 
If I try to remove the multiple return statements, I will be ended up in long if condition block, I though of avoid that.  thanks